Fixes #18760 - Allow export to CSV

tbrisker opened the pull request:
This introduces a way of exporting tables from the UI to CSV.
There are 3 steps to adding a CSV export to a table:

1. Add the CsvResponder concern to the relevant controller.
2. Add a `format.csv` block to the index controller action. This block
should include a call to `csv_response` with the wanted resources.
3. Add `csv_link` to the title actions. (Or wherever you want the export
button to appear)

By default, the csv will include all columns that exist for that model
in the database, except for the `created_at` and `updated_at` ones. To
override the default, add a method `csv_columns` to the controllers that
returns an array of columns. Columns can be actual column names in the
database, reflections, or any method that can be called on an instance
of the model.

This commit also includes an example of the feature implemented on the
Hosts index page.

tbrisker committed:
Fixes #18760 - Allow export to CSV

This introduces a way of exporting tables from the UI to CSV.
There are 3 steps to adding a CSV export to a table:

1. Add the CsvResponder concern to the relevant controller.
2. Add a `format.csv` block to the index controller action. This block
should include a call to `csv_response` with the wanted resources.
3. Add `csv_link` to the title actions. (Or wherever you want the export
button to appear)

By default, the csv will include all columns that exist for that model
in the database, except for the `created_at` and `updated_at` ones. To
override the default, add a method `csv_columns` to the controllers that
returns an array of columns. Columns can be actual column names in the
database, reflections, or any method that can be called on an instance
of the model.

This commit also includes an example of the feature implemented on the
Hosts index page.

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -60,6 +60,9 @@ def index(title = nil)
       format.yaml { render :text => search.all(:select => "hosts.name").map(&:name).to_yaml }
       format.json
+      format.csv do
+        send_data search.to_csv(columns: [:id, :name, :os, :environment, :model, :hostgroup, :last_report])
```
can we add all attributes by default? I think most users won't mind to remove data vs adding would be more complex?

ohadlevy replied:
or maybe a way to reuse default json attributes?

ohadlevy replied:
also, should it export to a file by default? e.g.
```ruby
format.csv { send_data @users.to_csv, filename: "users-#{Date.today}.csv" }
```

tbrisker replied:
The default if no columns are passed is to use `column_names` which just gets all the columns from the database. Here I tried to match the UI table, but we don't have to. However `column_names` doesn't provide stuff like `hostgroup`, only `hostgroup_id` - this way we can call methods besides the data that is stored in the db.
I can add the filename export, sure.

ohadlevy commented on app/models/concerns/csv_exporter.rb:
```diff
@@ -0,0 +1,17 @@
+    def to_csv(columns: column_names, csv_options: {})
+      CSV.generate(csv_options) do |csv|
+        csv << columns.map(&:to_s)
+        columns.map!(&:to_sym)
```
why are you looping columns twice?

tbrisker replied:
`column_names` returns a list of strings, I want to be able to pass a list of symbols as well, and I need to make sure they are in the correct format in both places - headers are strings, sent parameters are symbols.

ohadlevy commented on app/models/concerns/csv_exporter.rb:
```diff
@@ -0,0 +1,17 @@
+      CSV.generate(csv_options) do |csv|
+        csv << columns.map(&:to_s)
+        columns.map!(&:to_sym)
+        all.limit(nil).find_each do |obj|
```
should consider using batches instead of loading all objects into memory.

ohadlevy replied:
really missing search filter here.

tbrisker replied:
`find_each` loads them in batches of 1000 by default.
Also, this method is called on a scope that includes the search filter.

ohadlevy commented on app/models/concerns/csv_exporter.rb:
```diff
@@ -0,0 +1,17 @@
+  module ClassMethods
+    def to_csv(columns: column_names, csv_options: {})
+      CSV.generate(csv_options) do |csv|
+        csv << columns.map(&:to_s)
```
column_names is unused?

tbrisker replied:
`column_names` is an activerecord method that gives a list of all the column names, and it is used to provide the default value for the `columns` kwarg if no other value is passed.

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -60,6 +60,10 @@ def index(title = nil)
       format.json
+      format.csv do
+        send_data search.to_csv(columns: [:id, :name, :os, :environment, :model, :hostgroup, :last_report]),
+                  filename: "hosts-#{Date.today}.csv"
```
this generates N+1 Queries, see:
```
16:04:27 rails.1   | 2017-02-12T16:04:27 d703ba15 [app] [I] Sent data hosts-2017-02-12.csv (5.7ms)
16:04:27 rails.1   | 2017-02-12T16:04:27 d703ba15 [app] [I] Completed 200 OK in 408666ms (Views: 5.5ms | ActiveRecord: 4177.0ms)
16:04:27 rails.1   | 2017-02-12T16:04:27 d703ba15 [app] [W] user: ohad
16:04:27 rails.1   |  | GET /hosts.csv
16:04:27 rails.1   |  | USE eager loading detected
16:04:27 rails.1   |  |   Host::Managed => [:operatingsystem]
16:04:27 rails.1   |  |   Add to your finder: :includes => [:operatingsystem]
16:04:27 rails.1   |  | Call stack
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `block (3 levels) in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `map'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `block (2 levels) in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:11:in `block in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:8:in `to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/hosts_controller.rb:64:in `block (2 levels) in index'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/hosts_controller.rb:51:in `index'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/concerns/application_shared.rb:14:in `set_timezone'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/lib/middleware/catch_json_parse_errors.rb:8:in `call'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/lib/middleware/tagged_logging.rb:18:in `call'
16:04:27 rails.1   |  | 
16:04:27 rails.1   |  | 
16:04:27 rails.1   | 2017-02-12T16:04:27 d703ba15 [app] [W] user: ohad
16:04:27 rails.1   |  | GET /hosts.csv
16:04:27 rails.1   |  | USE eager loading detected
16:04:27 rails.1   |  |   Host::Managed => [:environment]
16:04:27 rails.1   |  |   Add to your finder: :includes => [:environment]
16:04:27 rails.1   |  | Call stack
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `block (3 levels) in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `map'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `block (2 levels) in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:11:in `block in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:8:in `to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/hosts_controller.rb:64:in `block (2 levels) in index'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/hosts_controller.rb:51:in `index'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/concerns/application_shared.rb:14:in `set_timezone'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/lib/middleware/catch_json_parse_errors.rb:8:in `call'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/lib/middleware/tagged_logging.rb:18:in `call'
16:04:27 rails.1   |  | 
16:04:27 rails.1   |  | 
16:04:27 rails.1   | 2017-02-12T16:04:27 d703ba15 [app] [W] user: ohad
16:04:27 rails.1   |  | GET /hosts.csv
16:04:27 rails.1   |  | USE eager loading detected
16:04:27 rails.1   |  |   Host::Managed => [:model]
16:04:27 rails.1   |  |   Add to your finder: :includes => [:model]
16:04:27 rails.1   |  | Call stack
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `block (3 levels) in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `map'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `block (2 levels) in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:11:in `block in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:8:in `to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/hosts_controller.rb:64:in `block (2 levels) in index'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/hosts_controller.rb:51:in `index'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/concerns/application_shared.rb:14:in `set_timezone'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/lib/middleware/catch_json_parse_errors.rb:8:in `call'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/lib/middleware/tagged_logging.rb:18:in `call'
16:04:27 rails.1   |  | 
16:04:27 rails.1   |  | 
16:04:27 rails.1   | 2017-02-12T16:04:27 d703ba15 [app] [W] user: ohad
16:04:27 rails.1   |  | GET /hosts.csv
16:04:27 rails.1   |  | USE eager loading detected
16:04:27 rails.1   |  |   Host::Managed => [:hostgroup]
16:04:27 rails.1   |  |   Add to your finder: :includes => [:hostgroup]
16:04:27 rails.1   |  | Call stack
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `block (3 levels) in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `map'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:12:in `block (2 levels) in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:11:in `block in to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/csv_exporter.rb:8:in `to_csv'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/hosts_controller.rb:64:in `block (2 levels) in index'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/hosts_controller.rb:51:in `index'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/controllers/concerns/application_shared.rb:14:in `set_timezone'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/lib/middleware/catch_json_parse_errors.rb:8:in `call'
16:04:27 rails.1   |  |   /home/ohad/git/foreman/lib/middleware/tagged_logging.rb:18:in `call'
16:04:27 rails.1   |  | 
```

tbrisker commented on app/controllers/hosts_controller.rb:
```diff
@@ -48,18 +48,20 @@ def index(title = nil)
         render :index if title && (@title = title)
       end
-      format.yaml { render :text => search.all(:select => "hosts.name").map(&:name).to_yaml }
-      format.json
```
These formats seem to be broken and not used anywhere anyways afaict.

tbrisker commented on app/controllers/templates_controller.rb:
```diff
@@ -99,6 +99,10 @@ def export
     send_data @template.to_erb, :type => 'text/plain', :disposition => 'attachment', :filename => @template.filename
   end
 
+  def resource_class
```
need resource_class to not be private so the index view can check if the to_csv method is supported.

tbrisker replied:
actually this is no longer necessary, but i think it makes sense to leave it here anyways for consistency with the default.

ohadlevy commented:
@tbrisker set WoC to handle N+1 problem. thanks.

tbrisker commented:
@ohadlevy I think I fixed it in my last push, could you please check again?

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -48,18 +48,20 @@ def index(title = nil)
       error e.to_s
       search = resource_base.search_for ''
     end
+    @hosts = search.includes(included_associations).paginate(:page => params[:page])
```
testing with newer version is a lot faster:
```
17:12:27 rails.1   | 2017-02-12T17:12:27 c58ea131 [app] [I] Sent data hosts-2017-02-12.csv (0.5ms)
17:12:27 rails.1   | 2017-02-12T17:12:27 c58ea131 [app] [I] Completed 200 OK in 61069ms (Views: 0.4ms | ActiveRecord: 1690.6ms)
```

However, most of the AR time was spent in Token and Host Status loading, both are not reflected in the CSV, in that regard, it does open the question if we should enable power status and host-status visibility or not in the csv.

tbrisker replied:
I think power status would require much too long for it to be useful as it will have to query external apis. Not sure about status, might also be complicated to include as it is computed currently in a complex way and not as a method on host that we can easily call. I'll remove for now and see how that goes.

tbrisker commented:
@ohadlevy removed power and status related includes, please try again.

thomasmckay commented:
This is a fairly significant feature that will impact all the tables. Is it WYSIWYG? Do the other export tools (hammer csv, rake tasks, etc.) need to be updated to leverage this new functionality? Are plugin data extensions taken into account?

ohadlevy reviewed (changes requested):
@tbrisker I would still want to see two more items in this PR:

1. the ability to pass the attributes in which a user can ask for data, for example:
/hosts.csv?search=<>&fields=<fieldA,fieldB..>

2. extract the default fields into a method, so plugins can override it easily.

ohadlevy commented:
@rohoover what do you think about the icon and placement of the action(note its a img while hovering the btn)? thanks!

![image](https://cloud.githubusercontent.com/assets/20167/22874834/f84c52cc-f1d0-11e6-879e-31854e0e3b1f.png)

dLobatog commented:
@ohadlevy To me that image looks like I'm exporting the results of the search, instead of whatever is actually shown in the table. I would say next to the pagination numbers it would convey the meaning of 'export this table' better - but I defer to @rohoover - might be just a first impression

Rohoover commented:
@ohadlevy @dLobatog 

If it is exporting the table results, then I would not put it on the search bar. I would place it with the table actions. That is where many of the export buttons in the application are located. I'd also compare the other export buttons for verbiage and iconography. I do not think all the export buttons have an icon.

ohadlevy commented:
@tbrisker latest test shows another big improvement!
```
17:58:18 rails.1   | 2017-02-13T17:58:18 d617a551 [app] [I] Sent data hosts-2017-02-13.csv (13.8ms)
17:58:18 rails.1   | 2017-02-13T17:58:18 d617a551 [app] [I] Completed 200 OK in 42063ms (Views: 13.6ms | ActiveRecord: 645.7ms)
```

sadly you can still see that out of the 42 seconds, only 6.5 seconds were in the db, the rest is mostly in the csv generation.

lzap commented on app/controllers/hosts_controller.rb:
```diff
@@ -58,8 +58,11 @@ def index(title = nil)
+      format.csv do
+        @hosts = search.includes(included_associations - [:host_statuses, :token, :compute_resource])
+        send_data @hosts.to_csv(columns: [:id, :name, :os, :environment, :model, :hostgroup, :last_report]),
+                  filename: "hosts-#{Date.today}.csv"
```
One note, this will easily eat all memory of a server process for big tables like reports or audits. While `send_data` streams (which is exactly what we want to do), method `to_csv` creates in-memory string. We already treat memory badly in Foreman and we do not want to make it worse, so my point is, can you rewrite this into a helper method that will stream the data row by row. Ideally you can read the rows in batches using `find_in_batches` so it's also "streaming" from DB directly.

lzap reviewed (changes requested):
Requesting small change to avoid allocating in-memory string completely.

ohadlevy commented:
@lzap something like http://smsohan.com/blog/2013/05/09/genereating-and-streaming-potentially-large-csv-files-using-ruby-on-rails/ ?

lzap commented:
> @lzap something like http://smsohan.com/blog/2013/05/09/genereating-and-streaming-potentially-large-csv-files-using-ruby-on-rails/ ?

Yes, something like that. I think enterprise users will totally love this, we will create API for that as well, the code will be reused, they will run it every hour and bang - memory usage goes wild.

ohadlevy commented:
@Rohoover do you have a suggestion of where and what text/icon to place?

Rohoover commented:
@ohadlevy 

Probably to the right of the table actions... in the screen I see a Select All. In that row, most likely not the first action. If the CSV file format is common, then I would just have a button that says "Export". If CSV is uncommon, the "Export CSV"

dLobatog commented:
@Rohoover The button you're seeing is not 'Select all' but 'Select Action', which looks like this already:

![screenshot from 2017-02-16 09-59-05](https://cloud.githubusercontent.com/assets/598891/23014282/93cdbfa6-f42e-11e6-8335-95237cff5399.png)

I thing the button is overused, and it represents "actions to be applied to the hosts". The option I was thinking of, is to put the icon next to the "Displaying all entries (bottom left corner)" or next to the pagination numbers like in this table:

![screenshot from 2017-02-16 09-58-43](https://cloud.githubusercontent.com/assets/598891/23014298/b8cbdcca-f42e-11e6-94f7-b657e39d9a70.png)

ShimShtein commented:
@ohadlevy, about:

> the ability to pass the attributes in which a user can ask for data, for example:
/hosts.csv?search=<>&fields=<fieldA,fieldB..>

Do we want to consider a some standard way of specifying queries? I would suggest [JSON API](http://jsonapi.org/format/#fetching-sparse-fieldsets), since we could benefit from using this standard in other areas too (I am talking about nested fields update semantics as opposed to full overwrite that we have today).

Rohoover commented:
@dLobatog That is the longest list of actions I've ever seen. Yikes!

I am still of the opinion that an action should not be placed with the pagination or total display entry field. I've been hunting around and that would be a highly unusual interaction pattern. For an action affecting the table or multiple rows, it would be placed in the same line with "Select Action", "Select Host". I would create a new button in that row for Export.

tbrisker commented:
Updated the PR:

* Stream data instead of building whole csv in memory
* Move button to title actions and change from icon to "Export".
* Changed CSV Exporter from an activerecord concern to a service.
* Added CSV Builder controller concern to set headers and initiate the data streaming. 
* Extract default fields to a method in the concern that can be overridden in the controller. 

Not in this PR: Allowing users to specify columns, as this requires checking user input for permitted values (otherwise a user can ask for a column `delete` which will delete the entire table for example). I would rather separate that ability to a future PR so this one doesn't grow too big to review (which is also the reason I added the button only on one table).

tbrisker commented:
setting to WoC since I need to fix the tests as well.

tbrisker commented:
@ohadlevy @lzap Care to have another look? tests have been fixed.

dLobatog commented on app/services/csv_exporter.rb:
```diff
@@ -0,0 +1,21 @@
+    Enumerator.new do |csv|
+      csv << csv_header(columns)
+
+      resources.find_each do |obj|
```
I've ran into limitations with `find_each` on trying to use this with objects in Katello API controllers. The issue is, find_each is OK when you want to call it on ActiveRecord relations. However some controllers directly return Arrays/Hashes etc.. 

I will check what to do with those so that those controllers can work with this.

dLobatog commented on app/services/csv_exporter.rb:
```diff
@@ -0,0 +1,21 @@
+      csv << csv_header(columns)
+
+      resources.find_each do |obj|
+        csv << csv_row(obj, columns)
```
I think `obj` here has to be flattened. Otherwise relations (e.g: column_names for Product has `sync_plan`, but `sync_plan` by itself will just return an object) will not show up.

If it's not flattened, we will have to write a `to_s` for everything that doesn't display well as a string

dLobatog replied:
That example actually isn't good. Feel free to disregard until I provide better feedback.

tbrisker commented:
@ohadlevy @lzap Fixed streaming, should work now. It was missing a header before (`no-cache`) which strangely didn't work for me yesterday. Also changed a couple of names to be more expressive, and removed a function call in the main csv loop (didn't see a significant difference in timing for 40K records but might be significant in a larger dataset or in memory consumption).

lzap commented:
Looks good to me, thanks for digging into streaming so memory does not goes high. Kudos!

thomasmckay commented:
See https://github.com/Katello/katello/pull/6646#issuecomment-284859529

Would it be possible to change the UI button to be a dropdown of choices in this PR? Follow up PRs could then add choices to match the existing CSV exports (hammer --csv host list and hammer csv hosts --export).

ohadlevy commented:
test foreman

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -58,8 +59,10 @@ def index(title = nil)
-      format.yaml { render :text => search.all(:select => "hosts.name").map(&:name).to_yaml }
-      format.json
+      format.csv do
+        @hosts = search.includes(included_associations - [:host_statuses, :token, :compute_resource])
```
I think its harmless, but in the logs i see:
```
  [sql] [W] Scoped order and limit are ignored, it's forced to be batch order and batch size
```

tbrisker replied:
I could clear the order and limit explicitly to get rid of the warning.

ohadlevy reviewed (approved):
Streaming is awesome, good job @tbrisker

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -918,4 +921,8 @@ def host_attributes_for_templates(host)
   end
+
+  def csv_columns
+    [:id, :name, :operatingsystem, :environment, :model, :hostgroup, :last_report]
```
What do you think about removing the ID column? I'm not sure if its required (nor its visible in the current table).

tbrisker replied:
I'm not sure, might be useful to have it just in case but I don't mind removing either.

ohadlevy replied:
I would drop the id.

ohadlevy commented:
@tbrisker please note the test failures.

tbrisker commented:
@ohadlevy Pushed again with cleared limit and order. previous test failures seem unrelated, let's see what happens now.

tbrisker commented:
@thomasmckay You are welcome to add or ask for more features to this, but please in a separate PR. 
As I mentioned already, this is not meant to replace or be identical in output to other CSV tools, this feature only tries to address the "I want this table I see on the screen in front of me in csv format" use case. 
Future additions can include allowing users to select which columns they want (http://projects.theforeman.org/issues/18761). 
If there are other use cases or features you want to add, feel free to add them to the tracker (http://projects.theforeman.org/issues/18463).

ohadlevy reviewed (changes requested):
@tbrisker :+1: for the code, I would like to add a bit more testing, if you can:
1. add tests for controller headers
2. add tests that you actually get enum

also following on nested / ref data, maybe we should have a test that fails csv response if it has _id in an attribute name?

tbrisker commented:
@ohadlevy Added controller test, dropped ID.

ohadlevy commented:
pending Mr. Jenkins.

ohadlevy commented:
thanks @tbrisker
